### DISCUSSION TITLE
Unable to connect to device (Bluetooth RGBW bulb through gateway)

### DISCUSSION THREAD

**Discussion by felipecrs** [make-all/tuya-local#1425] [Q&A]
When I try to add my Bluetooth RGB bulb in Tuya Local through a gateway, the following happens:

![image](https://github.com/make-all/tuya-local/assets/29582865/edb433b9-ab8a-4077-8afc-94bd23b1552c)

I thought I had to open an issue to add support for a new device, however, I don't receive any helpful logs in my integration (and the issue template asks me to open a discussion in such case instead). This is what I found in my logs:

```
2023-12-18 10:38:57.637 WARNING (MainThread) [homeassistant.util.async_] Detected blocking call to sleep inside the event loop by custom integration 'tuya_local' at custom_components/tuya_local/device.py, line 74: parent=tinytuya.Device(dev_id, address, local_key),, please create a bug report at https://github.com/make-all/tuya-local/issues
```

So, I'm opening this discussion to get help as to what can be done about it.

Thanks for the awesome integration.

PS: initial support for such bulb was asked in https://github.com/make-all/tuya-local/issues/275.

**Comment by robertrebor** [comment]
I have exactly same problem as you. I received complet information over tinytuya, but I am not able to add light to integration. From tinytua I have:
 "name": "Light 2",
        "id": "bfe135gqXXX",
        "key": "78ffa7a11fd7XXXX",
        "mac": "DC:23:50:04:12:8E",
        "uuid": "uuidde70b98XXX",
        "category": "dj",
        "product_name": "SD TB GU10 ",
        "product_id": "8kfokmpq",
        "biz_type": 18,
        "model": "220V\u84dd\u7259\u706f\u676f",
        "sub": true,
        "icon": "https://images.tuyaeu.com/smart/icon/bay1638666472267gXxe/6d54ae0008dd4e575e949cd09d7f7f9b.png",
        "node_id": "dc2350041XXX",
        "mapping": {
            "1": {
                "code": "switch_led",
                "type": "Boolean",
                "values": {}
            },
            "2": {
                "code": "work_mode",
                "type": "Enum",
                "values": {
                    "range": [
                        "white",
                        "colour",
                        "scene",
                        "music"
                    ]
                }
            },
            "3": {
                "code": "bright_value_v2",
                "type": "Integer",
                "values": {
                    "min": 10,
                    "max": 1000,
                    "scale": 0,
                    "step": 1
                }
            },
            "4": {
                "code": "temp_value_v2",
                "type": "Integer",
                "values": {
                    "min": 0,
                    "max": 1000,
                    "scale": 0,
                    "step": 1
                }
            },
            "7": {
                "code": "countdown_1",
                "type": "Integer",
                "values": {
                    "unit": "s",
                    "min": 0,
                    "max": 86400,
                    "scale": 0,
                    "step": 1
                }
            },
            "34": {
                "code": "do_not_disturb",
                "type": "Boolean",
                "values": {}
            }
        },
        "parent": ""
    },
    {

**Comment by robertrebor** [comment]
I have additional information, I checked official tuya integration in HA. I found that my Light 2 is loaded with integration, but has no entities. It means problem don't will be in tuya local, but something with bluetooth lights connected over hub.

<img width="442" alt="image" src="https://github.com/make-all/tuya-local/assets/114231763/aada05eb-50ac-444e-9f9d-317b316dd8f9">

**Comment by felipecrs** [comment]
@make-all is there a chance you can provide some support here? It's still behaving the exact same way as today.

One theory I have is that this device doesn't send any information back. It only allows to send commands to it, but you can't read anything from it.

**Comment by almirus** [comment]
@felipecrs How many sub devices do you have forwarded through the hub?  there is a limit on the number
https://github.com/make-all/tuya-local/discussions/1581

**Reply by felipecrs** [reply]
Configured in the Tuya Local integration? None. But in the Tuya app, there are three subdevices registered in the hub. All the same light bulb model.

**Reply by almirus** [reply]
This is your problem 😌

**Reply by felipecrs** [reply]
Oh, really? Do you think I should give https://github.com/grayyang/tuya-local a try?

**Reply by felipecrs** [reply]
Would unregistering 2 of my lights from my Tuya app also work? It's a little easier than testing that fork, since it's outdated.

**Reply by almirus** [reply]
Through this integration, based on my observations, you can connect up to 3 identical sub-devices. if you delete the lamps from the app, you will not be able to add them to the integration (the access keys will become invalid).

**Reply by felipecrs** [reply]
I'm confused now. First I said I had 3 identical sub-devices and you said that was my problem, now you say it's working for you with 3 identical sub-devices?

**Reply by almirus** [reply]
Are there other connected devices through the hub already? Is it the first bulb that isn’t connecting, or the third one (the first two connected normally)? Perhaps I’m just confused 🤣 – out of my six bulbs, only three have connected. The link I provided says it’s an integration issue

**Reply by felipecrs** [reply]
I can't connect any bulb with Tuya Local through my hub. They work normally in the Tuya App though (through my hub).

**Reply by almirus** [reply]
Are there **other** connected devices through the hub already in tuya-local?

**Reply by felipecrs** [reply]
No, there are none.

**Reply by almirus** [reply]
Then I apologize. What I wrote is relevant when several identical devices are connected via a hub and this integration. In your case, either there is simply no support for your device yet, or the protocol was chosen incorrectly, or there are errors in the integration.

**Reply by felipecrs** [reply]
Alright, no problem.

I am suspecting the integration is trying to fetch information from my light bulb like status on or off, but my bulb doesn't support it. It only supports receiving commands, not sending anything back.

**Reply by make-all** [reply]
It depends on the hub, some only allow one connection, some allow 3.  But since this is happening with the first device, it may not be related to the problem that @greylang has been working on.

Another possible issue is that sub devices seem to use the uuid or node_id rather than the device id.  I don't have any myself, so I cannot confirm what is the correct field to use.

**Reply by felipecrs** [reply]
Hm... just tried with uuid here rather than node_id, but unfortunately same issue.

```
Registrador: custom_components.tuya_local.device
Fonte: custom_components/tuya_local/device.py:557
integração: Tuya Local (documentação, problemas)
Ocorreu pela primeira vez: 19:19:07 (1 ocorrências)
Último registro: 19:19:07

Failed to refresh device state for Test.
```

```json
{
        "name": "Luz do escrit\u00f3rio 1",
        "id": "ebdad51ecb8ae61307ygjm",
        "key": "hidden?",
        "mac": "DC:23:4D:C5:9E:7A",
        "uuid": "e4c6a27fe358b096",
        "sn": "2000017900635E",
        "category": "dj",
        "product_name": "Beacon RGBCW",
        "product_id": "eyipa0uv",
        "biz_type": 0,
        "model": "A70",
        "sub": true,
        "icon": "https://images.tuyaus.com/smart/icon/bay1615007314201xIxe/f7b7ea380a52af57f098d87baea6252b.png",
        "node_id": "dc234dc59e7a",
        "mapping": {
            "1": {
                "code": "switch_led",
                "type": "Boolean",
                "values": {}
            },
            "2": {
                "code": "work_mode",
                "type": "Enum",
                "values": {
                    "range": [
                        "white",
                        "colour",
                        "scene",
                        "music"
                    ]
                }
            },
            "3": {
                "code": "bright_value_v2",
                "type": "Integer",
                "values": {
                    "min": 10,
                    "max": 1000,
                    "scale": 0,
                    "step": 1
                }
            },
            "4": {
                "code": "temp_value_v2",
                "type": "Integer",
                "values": {
                    "min": 0,
                    "max": 1000,
                    "scale": 0,
                    "step": 1
                }
            },
            "7": {
                "code": "countdown_1",
                "type": "Integer",
                "values": {
                    "unit": "s",
                    "min": 0,
                    "max": 86400,
                    "scale": 0,
                    "step": 1
                }
            },
            "34": {
                "code": "do_not_disturb",
                "type": "Boolean",
                "values": {}
            }
        },
        "parent": "",
        "ip": "",
        "version": ""
    },
```

**Reply by robertrebor** [reply]
@felipecrs Do you have official tuya integration in the HA? Can you check if in the official tuya integration do you have any entities for this device? I closed this topic from my side because I think that if in the official tuya integration are not entities provided, then tuya-local integration has no possibility to add this device.

**Reply by felipecrs** [reply]
Yes. The official Tuya integration for Home Assistant is able to control my device through my hub.
